Remove the FIFO thread

sakertooth opened the pull request:
Removes the FIFO thread, which was a thread used in the audio engine that wrote chunks of rendered audio buffers to a circular buffer, while the audio callback thread would read from that. Instead, the audio callback is the one rendering the buffers directly and doing any chunking as necessary.

The functionality this thread brought allowed for buffers to be "written ahead" of the audio callback, meaning that the FIFO thread would be rendering buffers as fast it can, stopping when it can't write any more buffers, while the audio callback would be reading them at a fixed interval. However:
1. Since the priority of this thread was the same as every other thread running on the system (i.e., not high priority), while the audio callback is commonly thought to run on high priority, it is more likely that the FIFO thread wouldn't have gotten much time to do its task of rendering (i.e., a high priority task). Moving that work into the audio callback is better in this case.
2. Even if the FIFO thread had high priority, it would have to compete with the audio callback's high priority as well. If we assume that the FIFO thread had high priority, but just a little bit less than the audio callback, then for buffers like 64, 32, etc, the audio callback thread will be running so frequently that we run the risk of not being able to render any buffers at all since the FIFO thread may not have a large enough time slice.
3. The latency wasn't completely determined by the buffer size, but also by how long it took to get this thread running since it was the one rendering the buffers.
4. Very strange deadlocks happened because of this thread. For example, simplifying the thread communication in this case fixes #7320.

This PR also greatly simplifies the transmission of buffers between the engine and the audio devices. `AudioEngine::renderNextBuffer` has been changed to correctly process the chunks of audio returned by `AudioEngine::renderNextPeriod` (a new function), and also manages both interleaved and planar outputs. This removes a lot of repetitive code in each audio device that originally was handling their configuration on their own.

This PR also makes the `AudioAlsa` device use floating-point audio instead of int16.

sakertooth commented:
Changed the PR to still remove the FIFO thread but keep the chunking of the buffer size (to avoid problems with non sample accurate automation but to also fix the problems this thread has caused).

Rossmaxx commented:
This shoulda had higher priority. @LMMS/testers

Rossmaxx commented on src/core/AudioEngine.cpp:
```diff
@@ -205,18 +166,8 @@ void AudioEngine::initDevices()
 
 
-void AudioEngine::startProcessing(bool needsFifo)
+void AudioEngine::startProcessing()
```
Would it make sense to inline it to the header?  Now that 90% of the functionality is gone.

sakertooth replied:
Yeah, probably.

Rossmaxx commented on src/core/AudioEngine.cpp:
```diff
@@ -225,18 +176,7 @@ void AudioEngine::startProcessing(bool needsFifo)
 
```
Same here.

Rossmaxx replied:
I mean the stop process function

Rossmaxx commented on src/core/audio/AudioDevice.cpp:
```diff
@@ -46,49 +45,17 @@ AudioDevice::AudioDevice( const ch_cnt_t _channels, AudioEngine*  _audioEngine )
-
-
-void AudioDevice::processNextBuffer()
+void AudioDevice::startProcessing()
```
Is this now an empty function?

sakertooth replied:
It seems like I removed it entirely.

Rossmaxx reviewed (commented):
The fatloss on this is insane. I am not approving for now because I am doubtful if something might have gotten lost somewhere. I will approve after i get around to testing.

Also, don't understand what is now happening with the chunking mechanism. Is the audio thread processing in chunks of 256 and then outputting to the playback/export at the user defined buffer size? Or is there something I missed.

bratpeki commented:
> @LMMS/testers
Reporting!

I'm busy with managing #7477, #7444 and #7366.

Is this making the code more readable without reducing performance? Or is it a performance boosting PR? In any case, great stuff, I'll look at it when I'm done with these three PRs!

Rossmaxx commented:
This is supposed to solve the performance degradation on master compared to 1.2 and the rare case with deadlocks making a mess, along with the buffer size and automation performance, among other stuff. @sakertooth correct me if I'm wrong. If what i guess is right, this PR has the potential to straighten up a lot of the performance bugs.

bratpeki commented:
> Wait the tester role is only for peki? 

For the time being, yeah. I think I get rights to review PRs and whatnot, so it's probably not wise to give it away to anyone wanting to test, but I'm not entirely sure about if that's how it behaves, just speculation.

sakertooth commented:
> This is supposed to solve the performance degradation on master compared to 1.2 and the rare case with deadlocks making a mess, along with the buffer size and automation performance, among other stuff. @sakertooth correct me if I'm wrong. If what i guess is right, this PR has the potential to straighten up a lot of the performance bugs.

It's mostly to fix deadlock bugs involving this thread waiting forever to close. It's less of a performance benefit (though that could be possible) but more so to simplify the thread communication within the audio pipeline.

CorruptVoidSoul commented:
So I tested this PR.
Exporting works fine, no hearable differences.
Playback is still pretty much the same as on master, I made it play the performance intensive part of a song, my phone is still in pain, no noticeable gain or loss there.
I had two random crashes but I expect this to be a me problem because well, it's a phone and this mmpz is uhh... a bit tough sometimes.
Still on the "me problem" side, usually Equalizer takes a minute to load for the first time, in this PR it takes 30 seconds so it's better, not gonna complain.
Now back on playback, the playhead will desync itself from the actual sound after a huge lag spike, but that's a general performance problem, it doesn't have much to do with this PR I think.

Final feedback is this seems good, did Peki test that too ? There could be differences with a computer that can handle heavy workload.

sakertooth commented:
> Final feedback is this seems good, did Peki test that too ? There could be differences with a computer that can handle heavy workload.

I don't think Peki has tested this yet, but they're probably busy with other PRs so it's fine. Thank you for testing this for me :+1:

sakertooth commented:
> Also, don't understand what is now happening with the chunking mechanism. Is the audio thread processing in chunks of 256 and then outputting to the playback/export at the user defined buffer size? Or is there something I missed.

Sorry, just realized that I didn't answer this question @Rossmaxx. What happens is that the user selects their desired buffer size and audio device to use, and then when LMMS starts, the audio device thread starts, which periodically calls `renderNextBuffer`, which either returns the latest rendered chunk (256 sample frames per chunk), or calls the overloaded variant for `renderNextBuffer` to render any buffer size wanted (e.g. 1024 frames). All of the audio devices call the overloaded variant for `renderNextBuffer` to be able to render any desired buffer size for real-time playback.

There is one thread requesting data periodically. That's it. There are other worker threads that do the actual processing, but this is the only thread requesting data (if not that thread then the export thread). To compare this to before, there was an additional thread, but this was removed, as it was mostly unnecessarily and causing issues.

sakertooth committed:
Remove FIFO thread

bratpeki commented:
I could test this, if it's ready to be tested, or needs WIP testing. @sakertooth ?

sakertooth commented:
> I could test this, if it's ready to be tested, or needs WIP testing. @sakertooth ?
There were merge conflicts, so I'm still working on this.

sakertooth committed:
Fix selection of frames per period

sakertooth committed:
Add "is running" flag for audio devices

sakertooth committed:
Remove extra start/stop flags in audio devices

sakertooth committed:
Revert "Remove extra start/stop flags in audio devices"

This reverts commit febe43308fc6f8b5c10bafcc30b0f77f9665f0b4.

sakertooth committed:
Store separate frames per period size in audio device

The audio engine has a chunked size, while the audio device has the full buffer size specified in the config

sakertooth committed:
Add chunking

sakertooth committed:
Fix dummy sound output

sakertooth committed:
Specify nullptr explicitly

sakertooth committed:
Attempt to fix MINGW and MSVC builds

sakertooth committed:
Attempt to fix MINGW and MSVC builds

sakertooth committed:
Pass in size parameter to getNextBuffer, return bool instead

sakertooth commented:
Should be ready @bratpeki.

bratpeki commented:
Please request a review so I see it on the GitHub UI :)

bratpeki commented:
Okay, so, how do I test this?

bratpeki commented:
Alright, I'll use it for some time as if it were the real deal, make a few songs, and report anything I see!

sakertooth committed:
Remove processNextBuffer and move writeBuffer into AudioDevice

sakertooth committed:
Create AudioDevice::nextBuffer function to simplify translation of rendered buffers between devices

Quite buggy, will fix

sakertooth commented:
PR is currently broken, will be fixed when I make this PR ready for review again.

sakertooth committed:
Fix nextBuffer function

sakertooth committed:
Calculate the number of frames based on the length parameter in sdlAudioCallback

sakertooth committed:
Use SND_PCM_FORMAT_FLOAT in ALSA

sakertooth committed:
Remove assert

sakertooth committed:
Fix comparison warnings

sakertooth committed:
Pass in correct frame count when fetching buffers in PortAudio callback

sakertooth committed:
Use PA_SAMPLE_FLOAT32 in PulseAudio

sakertooth committed:
Use pa_stream_begin_write to avoid extra copies

sakertooth committed:
Fix soundio callback

sakertooth committed:
Do not ignore return type on write call in OSS

sakertooth committed:
Ensure to silence buffer if nextBuffer returns false for callback-based devices

sakertooth committed:
Use same signedness

sakertooth committed:
Fix more signedness issues

sakertooth commented:
Applied the fixes, this PR should be good now.

bratpeki commented:
Tested with commit [29a6829](https://github.com/LMMS/lmms/commit/29a6829c2e1cb98206b6843bbdfe07f86ac31e33), seems to be just fine. I even noticed a performance boost! I believe Rossmaxx mentioned something about fewer deadlocks.

Testing the latest commit.

qnebra commented:
One of latest commits completly messed up playback on SDL backend on Windows. Only left channel can be heard and right channel is completly silent.

sakertooth committed:
Fix a few issues

sakertooth commented:
> One of latest commits completly messed up playback on SDL backend on Windows. Only left channel can be heard and right channel is completly silent.

Thanks for the heads up. It should be fixed now.

bratpeki commented:
I got a "Bus error (Core dumped)" and can't reliably recreate it. This only happened the first time I tried exporting a project made at a "later" LMMS (MIDI slicing, etc). So it could be tied to that. I'd rebase this to upstream or, because this is a HUGE improvement, just merge and we can make a follow-up PR.

bratpeki commented:
@regulus79 @Rossmaxx please have a look and if all is well, merge

sakertooth commented:
> @regulus79 @Rossmaxx please have a look and if all is well, merge

I'm not sure I'm done with this yet.

bratpeki commented:
Follow-up PR if needed, then!

Rossmaxx commented:
I have been thinking of getting back to this, but I'm in the middle of exams. Will take some time.

sakertooth committed:
Split nextBuffer declaration into two (one for interleaved, one for planar)

sakertooth committed:
Update src/core/audio/AudioAlsa.cpp

Co-authored-by: szeli1 <143485814+szeli1@users.noreply.github.com>

bratpeki commented:
We can merge now?

messmerd commented on include/AudioDevice.h:
```diff
@@ -75,66 +75,52 @@ class AudioDevice
-	}
-
-	virtual void stopProcessing();
+	virtual void startProcessing() { m_running.test_and_set(std::memory_order_acquire); }
+	virtual void stopProcessing() { m_running.clear(std::memory_order_release); }
```
Instead of needing to remember to call `AudioDevice::startProcessing` and `AudioDevice::stopProcessing` from the overridden methods in the child classes, why not let these be non-virtual wrappers that call some new pure virtual methods that the child classes would implement?

For example:
```cpp
public:
	void startProcessing()
	{
		m_running.test_and_set(std::memory_order_acquire);
		startProcessingImpl();
	}

protected:
	virtual void startProcessingImpl() = 0;
```

sakertooth replied:
I like this idea.

sakertooth replied:
Though, I think `startProcessingImpl` can actually be made `private` here.

sakertooth commented:
Thanks for the review @messmerd, will try to go over it soon, just a bit busy.

sakertooth commented on include/AudioDummy.h:
```diff
@@ -80,29 +80,23 @@ class AudioDummy : public QThread, public AudioDevice
-		while( true )
+		while (true)
 		{
+			audioEngine()->renderNextBuffer();
```
Probably. It was after before I made the changes too.

sakertooth committed:
Use static_cast instead of reinterpret_cast

Co-authored-by: Dalton Messmer <messmer.dalton@gmail.com>

sakertooth commented on src/core/audio/AudioPulseAudio.cpp:
```diff
@@ -243,35 +250,23 @@ void AudioPulseAudio::run()
+	auto buf = static_cast<void*>(nullptr);
+	auto maxBufSizeInBytes = framesPerPeriod() * channels() * sizeof(float);
 
+	pa_stream_begin_write(m_s, &buf, &maxBufSizeInBytes);
```
Yes. `maxBufSizeInBytes` is modified and returns the number of bytes, which is then later used to calculate the number of samples and audio frames. The code initially calculates the max size in bytes, but never assumes that `maxBufSizeInBytes == framesPerPeriod() * channels() * sizeof(float);`

sakertooth replied:
And if `maxBufSizeInBytes` isn't modified? The code still works anyways.

sakertooth committed:
Use new audio buffer view classes

sakertooth committed:
Render next buffer after resetting timer

sakertooth commented on src/core/audio/AudioOss.cpp:
```diff
@@ -250,31 +252,26 @@ void AudioOss::startProcessing()
+		for (auto i = std::size_t{0}; i < buf.size(); ++i)
 		{
-			break;
+			pcmBuf[i] = static_cast<int16_t>(buf[i] * OUTPUT_SAMPLE_MULTIPLIER);
```
Yess, we have `AFMT_S16_NE` for OSS, so yeah it shouldn't be an issue.

sakertooth committed:
Remove endian handling within audio devices

Most devices (even low-level ones like ALSA and OSS) already provide sample formats that allow you to ignore endianness, with some even handling the conversions internally.

sakertooth committed:
Make start and stop functions non-virtual, add pure virtual impl functions

sakertooth committed:
Make startProcessing and stopProcessing non virtual

sakertooth committed:
Remove unused function that was added in here

messmerd commented on src/core/audio/AudioDevice.cpp:
```diff
@@ -45,53 +49,69 @@ AudioDevice::AudioDevice( const ch_cnt_t _channels, AudioEngine*  _audioEngine )
-	const fpp_t frames = getNextBuffer( m_buffer );
-	if (frames) { writeBuffer(m_buffer, frames); }
-	else
+	if (!m_running.test_and_set(std::memory_order_acquire))
```
Are you using `m_running` like a mutex? Why wasn't something like this needed in `AudioDevice` prior to this PR?

sakertooth committed:
Remove temp

sakertooth committed:
Add isRunning function

bratpeki commented:
Tested again, @rdrpenguin04. Everything works well!

sakertooth committed:
Merge remote-tracking branch 'upstream' into fix-resampling

sakertooth committed:
Fix build

sakertooth commented on src/core/audio/AudioOss.cpp:
```diff
@@ -237,44 +208,36 @@ QString AudioOss::probeDevice()
-	{
-		start( QThread::HighPriority );
-	}
+	start(QThread::HighPriority);
```
> Should startProcessingImpl only be called when not running?

> If there's any possibility startProcessing could be called twice in a row, it could be rewritten like this:

I don't think theres any good reason why we would need to call `startProcessing` twice in a row. If somehow it is called twice in a row, it's probably a bug, so the `assert` approach makes more sense here IMO.

sakertooth commented on src/core/audio/AudioDevice.cpp:
```diff
@@ -45,54 +49,66 @@ AudioDevice::AudioDevice( const ch_cnt_t _channels, AudioEngine*  _audioEngine )
+		const auto audioEngineFrame = m_audioEngineBuffer[m_audioEngineBufferIndex];
 
-	if (!b) { return 0; }
+		for (auto channel = 0; channel < dst.channels(); ++channel)
```
I think it might be overly complex to do that. In this case, switching on `dst.channels()` should be a very predictable branch since the value is not changing, so the performance impact should be negligible.

sakertooth committed:
Merge remote-tracking branch 'upstream' into revamp-buffers

sakertooth committed:
Simplify nextBuffer function with new AudioBufferView concept

sakertooth committed:
Simplify loop even further

sakertooth commented on src/core/audio/AudioPulseAudio.cpp:
```diff
@@ -229,12 +221,10 @@ void AudioPulseAudio::run()
-		const fpp_t fpp = audioEngine()->framesPerPeriod();
-		auto temp = new SampleFrame[fpp];
-		while( getNextBuffer( temp ) )
+		while (true)
```
Good catch, should be using `AudioDevice::isRunning` I believe.

sakertooth committed:
Check if device running in while loop

sakertooth committed:
Remove m_quit variable

messmerd commented on src/core/audio/AudioDevice.cpp:
```diff
@@ -69,25 +69,35 @@ template void AudioDevice::nextBuffer<InterleavedBufferView<float>>(InterleavedB
 void AudioDevice::nextBuffer(AudioBufferView<float> auto dst)
 {
-	for (auto frame = std::size_t{0}; frame < dst.frames(); ++frame)
+	auto dstAt = [&dst](ch_cnt_t channel, f_cnt_t frame) -> float& {
+		return decltype(dst)::Interleaved ? dst[frame][channel] : dst[channel][frame];
+	};
```
Why not this one? It's not a major addition

sakertooth committed:
Add sample function in buffer view classes

sakertooth committed:
Remove framesPerPeriod from AudioDevice

+ Remove output buffer variables

sakertooth committed:
Set the buffer size for the JACK backend

sakertooth committed:
Use framesPerAudioBuffer where appropriate

sakertooth committed:
Move template instantiations down the file

sakertooth committed:
Do not call stopProcessing in destructor, assert that device is stopped before destruction

+ Remove m_stopped variable in JACK

sakertooth committed:
Merge remote-tracking branch 'upstream' into revamp-buffers

sakertooth committed:
Set frames per audio buffer within engine

sakertooth committed:
Add new renderNextPeriod function, move use of nextBuffer to renderNextBuffer

messmerd commented on src/core/AudioEngine.cpp:
```diff
@@ -66,27 +66,29 @@ using LocklessListElement = LocklessList<PlayHandle*>::Element;
+	: m_renderOnly(renderOnly)
+	, m_framesPerAudioBuffer(std::clamp(ConfigManager::inst()->value("audioengine", "framesperaudiobuffer").toULong(),
+		  MINIMUM_BUFFER_SIZE, MAXIMUM_BUFFER_SIZE))
+	, m_framesPerPeriod(std::min(m_framesPerAudioBuffer, DEFAULT_BUFFER_SIZE))
```
I must have been thinking that this code would run more than once, but `AudioEngine` is a singleton created at startup. Nevermind!

sakertooth committed:
Change assert

sakertooth committed:
Merge remote-tracking branch 'upstream' into revamp-buffers

sakertooth committed:
Fix build

sakertooth committed:
Force period size to DEFAULT_BUFFER_SIZE

sakertooth committed:
Use AudioBufferView again

sakertooth committed:
Update documentation

sakertooth committed:
Remove unused variables in SDL device

sakertooth commented on src/core/AudioEngine.cpp:
```diff
@@ -1045,49 +994,7 @@ MidiClient * AudioEngine::tryMidiClients()
-	m_fifo->waitUntilRead();
-}
+template void AudioEngine::renderNextBuffer(InterleavedBufferView<float> dst);
+template void AudioEngine::renderNextBuffer(PlanarBufferView<float> dst);
```
> I'd rather you have two separate overloads: one for `InterleavedBufferView` and one for `PlanarBufferView`. 

This is what I did in the beginning but the duplication was bugging me. If we do not care about the duplication issue then yeah I can do two overloads.

> Doing the explicit template instantiation in the source file like this is pretty unusual and doesn't offer much of a benefit.

Keeps the header file clean, and allows me to put this in the `.cpp` file. It seems to have a benefit. What's wrong with it?

> I'd be worried about potential linker errors if it's used incorrectly.
I wouldn't worry about things that haven't transpired yet. It hasn't caused me much issue so far.

> Also, in the future once we've switched completely over to planar buffers, wouldn't this also be planar-only?

It depends if the switch to planar only includes switching the audio backends to use planar as well (which I guess it makes sense to do so). And once we switch we can easily remove these explicit template instantiations if we want to.

messmerd replied:
I think I'd prefer duplicate code over a function template implementation in the .cpp file, but if you prefer to keep it this way, then that's fine

sakertooth committed:
Split renderNextBuffer into two overloads

sakertooth committed:
Revert "Force period size to DEFAULT_BUFFER_SIZE"

sakertooth committed:
Update documenation

sakertooth committed:
Switch fpp_t and f_cnt_t to std::uint64_t

sakertooth committed:
Move frameCount calculation down

sakertooth committed:
Cast frames per audio buffer to fpp_t

sakertooth committed:
Revert "Switch fpp_t and f_cnt_t to std::uint64_t"

This reverts commit e01046fc9dd67b535f60eb40edbccdc5dc837b4c.

sakertooth commented on src/core/AudioEngine.cpp:
```diff
@@ -68,26 +67,28 @@ using LocklessListElement = LocklessList<PlayHandle*>::Element;
-	m_audioDevStartFailed( false ),
-	m_profiler(),
-	m_clearSignal(false)
+AudioEngine::AudioEngine(bool renderOnly)
```
Given that there is no real danger if they are not a multiple anymore, throwing for these cases I find is excessive. If they are not a multiple, when then `renderNextBuffer` completes, a fraction of a period will be waiting to be read at a later time, which is accounted for in the design.

sakertooth committed:
Merge remote-tracking branch 'upstream' into revamp-buffers

sakertooth committed:
Merge remote-tracking branch 'upstream' into revamp-buffers

sakertooth committed:
Change type of frame count in AudioFileDevice::writeBuffer

sakertooth committed:
Inline bufferView

sakertooth committed:
Revert changes to AudioFileMP3.h

sakertooth committed:
Remove extra whitespace

sakertooth committed:
Remove m_devMutex and move header include of AudioBufferView.h

sakertooth committed:
Remove DEFAULT_SAMPLE_RATE

bratpeki reviewed (approved):
Approved for code and testing! 🎉

sakertooth committed:
Use AudioBufferView again, use one implementation and make it private

messmerd commented on src/core/audio/AudioPulseAudio.cpp:
```diff
@@ -243,35 +214,27 @@ void AudioPulseAudio::run()
+	}
+	else
+	{
+		audioEngine()->renderNextBuffer({reinterpret_cast<float*>(buf), channels(), numFrames});
```
```suggestion
		audioEngine()->renderNextBuffer({static_cast<float*>(buf), channels(), numFrames});
```

messmerd commented on include/AudioEngine.h:
```diff
@@ -253,10 +246,34 @@ class LMMS_EXPORT AudioEngine : public QObject
+	 *
+	 * If @a dst has 1 channel, the audio periods are averaged to mono.
+	 * If @a dst has 2 channels, the audio periods are directly copied.
+	 * If @a dst has more than 2 channels, the stero channels are copied and the rest are zero-filled.
```
```suggestion
	 * If @a dst has more than 2 channels, the stereo channels are copied and the rest are zero-filled.
```

sakertooth committed:
Fix misspelling in documentation

Co-authored-by: Dalton Messmer <messmer.dalton@gmail.com>

sakertooth committed:
Remove redundant std::array::fill call

Co-authored-by: Dalton Messmer <messmer.dalton@gmail.com>

sakertooth committed:
Use static_cast instead of reinterpret_cast

Co-authored-by: Dalton Messmer <messmer.dalton@gmail.com>

sakertooth committed:
Return early if pa_stream_begin_write returns a nonzero value

Co-authored-by: Dalton Messmer <messmer.dalton@gmail.com>

sakertooth committed:
Remove iostream include

sakertooth committed:
Update documentation

sakertooth committed:
Do not set JACK buffer size

sakertooth committed:
Fix some bugs with AudioJack, add input processing back in

sakertooth committed:
Mention possibility of removing framesPerAudioBuffer in a later refactor

messmerd commented on include/AudioEngine.h:
```diff
@@ -253,10 +240,55 @@ class LMMS_EXPORT AudioEngine : public QObject
+	 *
+	 * Renders @ref renderNextAudioPeriod() "audio periods" into @a dst. If @a dst is not a multiple of the period
+	 * size, the remaining frames are partially rendered (an extra period may be rendered in such cases, which can
+	 * degradge performance).
```
```suggestion
	 * degrade performance).
```

messmerd commented on include/AudioEngine.h:
```diff
@@ -253,10 +240,55 @@ class LMMS_EXPORT AudioEngine : public QObject
+	/**
+	 * @returns The internal buffer size used by audio plugins and other processing done within the audio engine.
+	 * It's maximum value is @ref DEFAULT_BUFFER_SIZE, or equal to @ref framesPerAudioBuffer() if it is less than @ref
+	 * DEFAULT_BUFFER_SIZE.
```
```suggestion
	 * Its value is @ref DEFAULT_BUFFER_SIZE or @ref framesPerAudioBuffer() - whichever is lower.
```

sakertooth replied:
Would prefer a comma but this works.

messmerd reviewed (approved):
I haven't tested it, but the code seems fine.

sakertooth committed:
Fix spelling in docs

Co-authored-by: Dalton Messmer <messmer.dalton@gmail.com>

sakertooth committed:
Simplify wording in docs

Co-authored-by: Dalton Messmer <messmer.dalton@gmail.com>

messmerd commented:
Time to merge?

sakertooth commented:
> Time to merge?

That should be ok.

messmerd commented on include/AudioPortAudio.h:
There's a compiler warning on macOS:
```
In file included from /Users/runner/work/lmms/lmms/src/core/audio/AudioPortAudio.cpp:33:
/Users/runner/work/lmms/lmms/include/AudioPortAudio.h:92:14: warning: private field 'm_outBufPos' is not used [-Wunused-private-field]
   92 |         std::size_t m_outBufPos = 0;
      |                     ^
```

`m_outBufPos` can be removed now.

messmerd commented on src/core/audio/AudioJack.cpp:
This doesn't appear related to this PR, but there's another easy-to-fix compiler warning in this file:
```
/Users/runner/work/lmms/lmms/src/core/audio/AudioJack.cpp:482:26: warning: lambda capture 'this' is not used [-Wunused-lambda-capture]
  482 |         auto buildToolButton = [this](QWidget* parent, const QString& currentSelection, const std::vector<std::string>& names, const QString& filteredLMMSClientName)
      |                                 ^~~~
```

sakertooth committed:
Remove outBuf from AudioPortAudio

sakertooth committed:
Fix unused lambda capture warning
Normally wouldn't do this (out of scope), but is a small change and this file was already altered

sakertooth committed:
Comma ftw

sakertooth committed:
Change renderNextPeriod to return a span
